tez-runtime-internals/src/main/java/org/apache/tez/runtime/task/TaskReporter.java

```diff
@@ -262,6 +263,13 @@ private synchronized ResponseWrapper heartbeat(Collection<TezEvent> eventsArg) t
           sendCounters = true;
           prevCounterSendHeartbeatNum = nonOobHeartbeatCounter.get();
         }
+        try {
+          task.checkTaskLimits();
```

Comment:
Doing this in heartbeat can significantly increase its runtime. As in, even small increase of few milliseconds can have impact on cluster usage. You may want to run this every 10 seconds here or in some other way. This should ideally handle sorter spills and merges, in which case it can handle another ticket which was created on similar lines.